transformations/change_date_format/transformation.py

```diff
@@ -0,0 +1,164 @@
+import spacy
+import dateparser
+
+from babel.core import LOCALE_ALIASES
+from babel.dates import format_date
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+
+class DateFormatTransformation:
+    nlp = None
+
+    def __init__(self, max_output=1):
+        self.nlp = spacy.load("en_core_web_sm")
+        self.max_output = max_output
+
+        self.ymd_formats = ["short", "medium", "long"]
+        self.ym_formats = ["MMM Y", "MMMM Y", "MMM YY", "MMMM YY"]
+        self.md_formats = [
+            "d MMM",
+            "d MMMM",
+            "dd MMM",
+            "dd MMMM",
+            "MMM d",
+            "MMMM d",
+            "MMM dd",
+            "MMMM dd",
+        ]
+
+        self.locales = [
+            "en_AU",
+            "en_CA",
+            "en_IN",
+            "en_IE",
+            "en_MT",
+            "en_NZ",
+            "en_PH",
+            "en_SG",
+            "en_ZA",
+            "en_GB",
+            "en_US",
+        ]
+
+    def parse_date(self, text: str):
+        """Parse the text to extract the date components and return a datetime object."""
+
+        # By defualt the parser fills the missing values with current day"s values,
+        # hence, using boolean values to keep track of what info is present in the text.
+        date = None
+        has_year = False
+        has_month = False
+        has_day = False
+
+        # First check if the text contains all three parts of a date - Y, M, D.
+        date = dateparser.parse(
+            text, settings={"REQUIRE_PARTS": ["year", "month", "day"]}
+        )
+
+        if date is not None:
+            has_year = True
+            has_month = True
+            has_day = True
+        else:
+            # Check if text contains two parts - Y and M.
+            date = dateparser.parse(text, settings={"REQUIRE_PARTS": ["year", "month"]})
+            if date is not None:
+                has_year = True
+                has_month = True
+            else:
+                # Check if text contains two parts - M and D.
+                date = dateparser.parse(
+                    text, settings={"REQUIRE_PARTS": ["month", "day"]}
+                )
+                if date is not None:
+                    has_month = True
+                    has_day = True
+
+        return date, has_year, has_month, has_day
+
+    def transform(self, input_text: str):
+        doc = self.nlp(input_text)
+
+        for entity in doc.ents:
+            new_value = None
+
+            if entity.label_ == "DATE":
+                date, has_year, has_month, has_day = self.parse_date(entity.text)
+
+                if date:
+                    if has_year and has_month and has_day:
+                        new_value = format_date(
+                            date,
+                            format=random.choice(self.ymd_formats),
+                            locale=random.choice(self.locales),
+                        )
```

Comment:
Doesn't random.seed need to be called before every random.choice?

Reply:
Also, you could add more variety by passing format as “full”, “long”, “medium”, or “short”, besides a custom date/time pattern?

Reply:
The "short/medium/long" options are useful only if the text contains all three fields - year, month and day. If any of the fields is missing in the original text, then by default it will be filled with the current value in the replacement. For example, if a text contains "March 2020", then it would be replaced as "22 March 2020" or "23 March 2020" depending on whether the current day is 22 or 23. Similarly, something like "December 31" which doesn't have the year specified  will be replaced as "December 31, 2021". Since this transformation is only intended to modify the format of the date, and not the actual date itself, I used custom date patterns to match whatever the original text contains so as to not introduce incorrect information.